### ISSUE TITLE
FEATURE: Support ENUM naming for ENUM reuse.

### ISSUE THREAD

**Issue by tiblu** [sequelize/sequelize#2577]
**Intro**
I want to name my ENUM type so that I could use the same ENUM in several Models.

**The need**
Right now ENUM name is created enum_ + ModelNamePlural + _ + fieldName. Problem arises when I "extend" a Model and inherit ENUM field 2 differently named ENUMs are created in the DB thus I have to do casting to actually compare the 2 in the queries. 

**Proposal**
ENUM type to take in "name" as parameter. That is:

```
            level: {
                type: DataTypes.ENUM('myEnumName'),
                values: ['admin', 'edit', 'read', 'none'],
                comment: 'User membership level.'
            }
```

The generated name would be enum_ + MyEnumName.

**Comment by frogcjn** [comment]
Any progress for this feature?

**Comment by naservere** [comment]
Hi, any news on this?

**Comment by apsavin** [comment]
I think, it's better to use the following syntax:

```js
Sequelize.ENUM({
    name: 'myEnumName',
    values: ['one', 'two']
})
```
Because ENUM already [can](https://github.com/sequelize/sequelize/blob/master/lib/data-types.js#L766) accepts values as shown above.
But here is a problem: nobody will want to write values every time when ENUM reuse needed.
So, may be, it's better to come with something like custom type definition syntax:

```js
sequelize.defineType('myEnumName', {
  as: Sequelize.ENUM, // or Sequelize.RANGE, or nothing, if you don't need `AS` in your query
  values: [...], // optional, for ENUM
  ...
})
```

**Comment by mickhansen** [comment]
Might make more sense to simply allow reusing of the ENUM type (+ naming)

**Comment by apsavin** [comment]
It seems that it's not too hard to add naming to the ENUM type (for postgres at least). 

And it seems that it can be reused right now:

```js
const myEnum = Sequelize.ENUM({
    name: 'myEnumName',
    values: ['one', 'two']
});

sequelize.define('model', {
   one: {
      type: myEnum,
   },

   two: {
      type: myEnum,
   },
})

```

**Comment by naservere** [comment]
@apsavin @mickhansen Thanks, guys.

**Comment by felixfbecker** [comment]
+1 for this syntax
```js
const myEnum = Sequelize.ENUM({
    name: 'myEnumName',
    values: ['one', 'two']
})

sequelize.define('User', {something: myEnum})
```

**Comment by sturdynut** [comment]
Any traction on this?

Currently in a situation where we could really use this feature.

**Comment by tiblu** [comment]
I guess we need someone to PR this. It should not be that hard and Sequelize has very comprehensive automated tests so it's easy to test and verify that nothing else was broken.

**Comment by papb** [comment]
@sturdynut Would you be willing to submit a PR? That would be great!

**Comment by sturdynut** [comment]
@papb - I will do my best to do up a PR soon.

**Comment by sturdynut** [comment]
Just a follow up.  I am making progress on this, slowly but surely.

**Comment by Betree** [comment]
I'm currently using the following workaround to share an enum between two models, it seems to work:

**Migration**
```es6
await queryInterface.createTable('TransactionSettlements', {
	kind: {
		// Re-using the same ENUM than `Transactions` so that we don't have to maintain two of them
		type: '"public"."enum_Transactions_kind"',
		allowNull: false,
	},
	...
});
```

**Model**
```es6
class TransactionSettlement extends Model {
	kind: TransactionKind; // Typescript enum from `Transactions`
	...
}

TransactionSettlement.init(
	kind: {
  		// Re-using the same ENUM than `Transactions` so that we don't have to maintain two of them.
 	 	// We'll have a better way of doing this with https://github.com/sequelize/sequelize/issues/2577
  		type: '"public"."enum_Transactions_kind"',
  		allowNull: false,
	},
	...
}
```

**Comment by jthoward64** [comment]
Just from a cursory look (correct me if I'm wrong) this looks like as of V7 all that is needed is changing [the pgEnum* functions](https://github.com/sequelize/sequelize/blob/10332eabdff7ebbb7a11964358938d40b5aa7542/packages/core/src/dialects/postgres/query-generator.js#L653), the ENUM [toSql function](https://github.com/sequelize/sequelize/blob/10332eabdff7ebbb7a11964358938d40b5aa7542/packages/core/src/dialects/postgres/data-types.ts#L382) and the [constructor](https://github.com/sequelize/sequelize/blob/10332eabdff7ebbb7a11964358938d40b5aa7542/packages/core/src/dialects/abstract/data-types.ts#L2146)/[type](https://github.com/sequelize/sequelize/blob/10332eabdff7ebbb7a11964358938d40b5aa7542/packages/core/src/dialects/abstract/data-types.ts#L2125). Am I wrong?

**Comment by jthoward64** [comment]
Also the behavior/syntax I'm thinking (to be friendly with typescript and decorators) is something like the following (this is from a failed attempt at using custom types for this behavior)

```typescript
export class NamedEnum<
  Member extends string = string
> extends DataTypes.ENUM<Member> {
  name: string;

  constructor(options: EnumOptions<Member> & { name: string }) {
    super(options);
    this.name = options.name;
  }

  toSql() {
    const context = this.usageContext;
    if (context == null) throw new Error("No usage context");

    return context.sequelize.queryGenerator.quoteIdentifier(this.name);
  }
}

export const AnEnum: NamedEnum = new NamedEnum({
  name: "enum_an_enum",
  values: ["some", "values"],
});
```

Then you would simply pass `anEnum` in for the data type of an attribute

**Comment by ephys** [comment]
That's correct, we redesigned the enum type in V7 to make it easier to implement this feature. Combined with a recent change to add "if not exists" to the enum creation query, it should be possible to implement this with few changes

**Comment by RWOverdijk** [comment]
Any in-between solution for us v6 users?
